### PR TITLE
Introduce "Bundler.clean_env"

### PR DESCRIPTION
**Pull request by njam**
Via https://github.com/bundler/bundler/issues/4232
Which should return the environment as it was *before* bundler modified it.

Should this use "original_env" or "clean_env" or both?
> At one time, the difference was that one went up one layer of env, while the other removed all Bundler values. I'm not sure if that's still true today.

You can assign this ticket to me if you want.

### CODE REVIEW

**Comment by njam** [comment]
@indirect @segiddins I added `original_env` and `clean_env` (the latter with a `@deprecated` flag). I've also rewritten the "with_clean_env_spec" test, I think it's simpler now and covers more.

What do you think?

File: lib/bundler.rb [hunk 1]
```diff
@@ -430,5 +440,14 @@ def upgrade_lockfile
...
       lockfile.rmtree
     end
+
+    # @param env [Hash]
+    def with_env(env)
+      backup = ENV.to_hash
+      ENV.replace(env)
+      yield
+    ensure
+      ENV.replace(backup.to_hash)
```
Comment (segiddins) [inline lib/bundler.rb]: The `to_hash` here is unnecessary

**Comment by segiddins** [comment]
@RochesterinNYC want to review the spec changes?

File: spec/runtime/with_clean_env_spec.rb [hunk 2]
```diff
@@ -2,107 +2,104 @@
...
 
 describe "Bundler.with_env helpers" do
-  shared_examples_for "Bundler.with_*_env" do
-    it "should reset and restore the environment" do
-      gem_path = ENV["GEM_PATH"]
-      path = ENV["PATH"]
+  describe "Bundler.original_env" do
+    it "should return the PATH present before bundle was activated" do
+      gemfile ""
+      bundle "install --path vendor/bundle"
```
Comment (RochesterinNYC) [inline spec/runtime/with_clean_env_spec.rb]: These two lines seem to be present in each of the `it` blocks for `describe "Bundler.original_env"` and `describe "Bundler.clean_env"`:
```ruby
      gemfile ""
      bundle "install --path vendor/bundle"
```

They could be extracted into a single `before` block.

File: spec/runtime/with_clean_env_spec.rb [hunk 3]
```diff
@@ -2,107 +2,104 @@
...
+      code = "print Bundler.original_env['PATH']"
 
-      Bundler.with_clean_env do
-        expect(`echo $GEM_PATH`.strip).not_to eq(gem_path)
-        expect(`echo $PATH`.strip).not_to eq(path)
+      path = `getconf PATH`.strip + ":/foo"
```
Comment (RochesterinNYC) [inline spec/runtime/with_clean_env_spec.rb]: extract these (`code =` and `path =`) into `let` statements?

**Comment by RochesterinNYC** [comment]
Other than a bit of possible DRYing up and more utilization of `rspec` features, the specs LGTM.

**Comment by njam** [comment]
Thanks for the feedback!

I moved the "bundle install" code to a `before` block.
I'm not sure about `let`. To replace all assignments with `let` I would need to wrap each test in a `context` block where I would then define those values (code, path, result). Is it the way to go?

**Comment by RochesterinNYC** [comment]
@njam Nope, you're right. It's possible to re-arrange these around to use `let` blocks but it's fine as is/isn't necessary. :+1:

**Comment by segiddins** [comment]
@homu r+

**Pull request merged**